core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java

```diff
@@ -0,0 +1,136 @@
+import org.apache.calcite.sql.SqlIntervalQualifier;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import static org.apache.calcite.util.RelToSqlConverterUtil.unparseHiveTrim;
+
+/**
+ * A <code>SqlDialect</code> implementation for the StarRocks database.
+ */
+public class StarRocksSqlDialect extends SqlDialect {
+
+  public static final SqlDialect.Context DEFAULT_CONTEXT = SqlDialect.EMPTY_CONTEXT
+      .withDatabaseProduct(SqlDialect.DatabaseProduct.STARROCKS)
+      .withIdentifierQuoteString("`")
+      .withNullCollation(NullCollation.LOW);
+
+  public static final SqlDialect DEFAULT = new StarRocksSqlDialect(DEFAULT_CONTEXT);
+
+  /**
+   * Creates a StarRocksSqlDialect.
+   */
+  public StarRocksSqlDialect(Context context) {
+    super(context);
+  }
+
+
+  @Override public boolean supportsCharSet() {
+    return false;
+  }
+
+  @Override public void unparseOffsetFetch(SqlWriter writer, @Nullable SqlNode offset,
+      @Nullable SqlNode fetch) {
+    unparseFetchUsingLimit(writer, offset, fetch);
+  }
+
+  @Override public @Nullable SqlNode emulateNullDirection(SqlNode node,
+      boolean nullsFirst, boolean desc) {
+    return emulateNullDirectionWithIsNull(node, nullsFirst, desc);
+  }
+
+  @Override public boolean supportsApproxCountDistinct() {
+    return true;
+  }
+
+  @Override public boolean supportsNestedAggregations() {
+    return false;
+  }
+
+  @Override public boolean supportsTimestampPrecision() {
+    return false;
+  }
+
+  @Override public CalendarPolicy getCalendarPolicy() {
+    return CalendarPolicy.SHIFT;
+  }
+
+  @Override public void unparseCall(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    switch (call.getKind()) {
+    case ARRAY_VALUE_CONSTRUCTOR:
+      final SqlWriter.Frame arrayFrame = writer.startList("[", "]");
+      for (SqlNode operand : call.getOperandList()) {
+        writer.sep(",");
+        operand.unparse(writer, leftPrec, rightPrec);
+      }
+      writer.endList(arrayFrame);
+      break;
+    case MAP_VALUE_CONSTRUCTOR:
+      writer.keyword(call.getOperator().getName());
+      final SqlWriter.Frame mapFrame = writer.startList("{", "}");
+      for (int i = 0; i < call.operandCount(); i++) {
+        String sep = i % 2 == 0 ? "," : ":";
+        writer.sep(sep);
+        call.operand(i).unparse(writer, leftPrec, rightPrec);
+      }
+      writer.endList(mapFrame);
+      break;
+    case TRIM:
+      unparseHiveTrim(writer, call, leftPrec, rightPrec);
+      break;
+    case FLOOR:
+      MysqlSqlDialect.DEFAULT.unparseCall(writer, call, leftPrec, rightPrec);
+      break;
+    default:
+      super.unparseCall(writer, call, leftPrec, rightPrec);
+      break;
+    }
+  }
+
+  @Override public @Nullable SqlNode getCastSpec(RelDataType type) {
+    switch (type.getSqlTypeName()) {
+    case TIMESTAMP:
```

Comment:
I am confused what's going on here.
If the Calcite TIMESTAMP type is called "DATETIME" in StarRocks, I am not sure whether this is the only case that needs changing. Can you write some tests that use TIMESTAMP?

Reply:
Exactly,  Similar to MySQL, the Calcite TIMESTAMP type is called "DATETIME" in StarRocks.  This processing refers to `MysqlSqlDialect#getCastSpec`.

Reply:
Added additional TIMESTAMP test:
- `RelToSqlConverterTest#testCastToTimestamp`
- `RelToSqlConverterTest#testUnparseSqlIntervalQualifier`